### PR TITLE
Fix properties of template members and add signature as astValue.

### PR DESCRIPTION
**Pull request by bruntib**
If a template class has a method then its definition also has to
be visited so we can jump on it and get its mangled name.

Moreover for function we add the signature as astValue. We use
astValue for textual representation of an AST node. This way we don't
need to get the properties of a function just to get its signature.

### CODE REVIEW

File: plugins/cpp/service/src/diagram.cpp [hunk 1]
```diff
@@ -354,14 +351,12 @@ std::string Diagram::getDetailedClassNodeLabel(const AstNodeInfo& nodeInfo_)
...
 
-  props = getProperties(nodes);
-
   for (const AstNodeInfo& node : nodes)
   {
     std::string visibility = visibilityToHtml(node);
 
     // TODO: Constructor and Destructor signatures can be empty.
-    std::string signature = props[node.id]["Signature"];
+    std::string signature = getProperty(node.id, "Signature");
```
Comment (csordasmarton) [inline plugins/cpp/service/src/diagram.cpp]: You stored the signature in the astValue. Why didn't you use it?

File: plugins/cpp/service/src/diagram.cpp [hunk 2]
```diff
@@ -329,14 +329,11 @@ std::string Diagram::getDetailedClassNodeLabel(const AstNodeInfo& nodeInfo_)
...
 
-  std::map<core::AstNodeId, std::map<std::string, std::string>> props
-    = getProperties(nodes);
-
   for (auto it = nodes.begin(); it != nodes.end(); ++it)
   {
     std::string visibility = visibilityToHtml(*it);
     std::string content = memberContentToHtml(*it,
-      util::escapeHtml(it->astNodeValue + " : " + props[it->id]["Type"]));
+      util::escapeHtml(it->astNodeValue + " : " + getProperty(it->id, "Type")));
```
Comment (csordasmarton) [inline plugins/cpp/service/src/diagram.cpp]: Why did you change back the signature of the `getProperties` function? Now you do a lot of query which is not good for performance.

File: plugins/cpp/service/src/cppservice.cpp [hunk 3]
```diff
@@ -186,89 +186,81 @@ void CppServiceHandler::getAstNodeInfoByPosition(
 }
 
 void CppServiceHandler::getProperties(
-  std::map<core::AstNodeId, std::map<std::string, std::string>>& return_,
-  const std::vector<core::AstNodeId>& astNodeIds_)
+  std::map<std::string, std::string>& return_,
+  const core::AstNodeId& astNodeId_)
 {
   using PropMap = std::map<std::string, std::string>;
```
Comment (csordasmarton) [inline plugins/cpp/service/src/cppservice.cpp]: You don't use it in this function.

File: plugins/cpp/parser/src/clangastvisitor.h [hunk 4]
```diff
@@ -600,7 +600,7 @@ class ClangASTVisitor : public clang::RecursiveASTVisitor<ClangASTVisitor>
 
     cppFunction->astNodeId = astNode->id;
     cppFunction->mangledNameHash = astNode->mangledNameHash;
-    cppFunction->name = getSignature(fn_);
+    cppFunction->name = astNode->astValue;
```
Comment (csordasmarton) [inline plugins/cpp/parser/src/clangastvisitor.h]: Why do we store the signature multiple times: once in the `astValue` of the `CppAstNode` and second in the `name` attribute of the `CppFunction`.

File: plugins/cpp/service/src/diagram.cpp (hunk 2 above)
Comment (bruntib) [inline plugins/cpp/service/src/diagram.cpp]: Good point. As the function signature has been placed in the `astNodeValue` attribute, it was no longer necessary to query properties for multiple functions. That's why I also reverted the changes of `getProperties()` API function. However that's true that now this results as many transactions as many attributes a class has in the diagram. I'll figure something out for this.

File: plugins/cpp/service/src/diagram.cpp (hunk 1 above)
Comment (bruntib) [inline plugins/cpp/service/src/diagram.cpp]: Thanks, I fix it.

File: plugins/cpp/parser/src/clangastvisitor.h (hunk 4 above)
Comment (bruntib) [inline plugins/cpp/parser/src/clangastvisitor.h]: Good point, thanks.
